src/stenographer.c

```diff
@@ -0,0 +1,372 @@
+                "user_data pointer is NULL.");
+        return (0);
+    }
+
+    if (len <= 0)
+        return (len);
+
+    if ((st->stenographer_buffer_fill + len) >= st->stenographer_buffer_size)
+    {
+        char *temp;
+
+        temp = (char *) realloc (st->stenographer_buffer,
+                st->stenographer_buffer_fill + len + 1);
+        if (temp == NULL)
+        {
+            ERROR ("stenographer plugin: realloc failed.");
+            return (0);
+        }
+        st->stenographer_buffer = temp;
+        st->stenographer_buffer_size = st->stenographer_buffer_fill + len + 1;
+    }
+
+    memcpy (st->stenographer_buffer + st->stenographer_buffer_fill, (char *) buf, len);
+    st->stenographer_buffer_fill += len;
+    st->stenographer_buffer[st->stenographer_buffer_fill] = 0;
+
+    return (len);
+} /* int stenographer_curl_callback */
+
+/* Configuration handling functiions
+ * <Plugin stenographer>
+ *   <Instance "instance_name">
+ *     URL ...
+ *   </Instance>
+ *   URL ...
+ * </Plugin>
+ */
+static int config_add (oconfig_item_t *ci)
+{
+    stenographer_t *st;
+    int i;
+    int status;
+
+    st = malloc (sizeof (*st));
+    if (st == NULL)
+    {
+        ERROR ("stenographer plugin: malloc failed.");
+        return (-1);
+    }
+    memset (st, 0, sizeof (*st));
+
+    status = cf_util_get_string (ci, &st->name);
+    if (status != 0)
+    {
+        sfree (st);
+        return (status);
+    }
+    assert (st->name != NULL);
+
+    for (i = 0; i < ci->children_num; i++)
+    {   
+        oconfig_item_t *child = ci->children + i; 
+        
+        if (strcasecmp ("URL", child->key) == 0)
+            status = cf_util_get_string (child, &st->url);
+        else if (strcasecmp ("Host", child->key) == 0)
+            status = cf_util_get_string (child, &st->host);
+        else if (strcasecmp ("Cert", child->key) == 0)
+            status = cf_util_get_string (child, &st->cert);
+        else if (strcasecmp ("Key", child->key) == 0)
+            status = cf_util_get_string (child, &st->key);
+        else if (strcasecmp ("CACert", child->key) == 0)
+            status = cf_util_get_string (child, &st->cacert);
+        else        
+        {   
+            WARNING ("stenographer plugin: Option `%s' not allowed here.",
+                    child->key);
+            status = -1;
+        }   
+    
+        if (status != 0)
+            break;
+    } 
+
+    /* Check if struct is complete.. */
+    if ((status == 0) && (st->url == NULL))
+    {
+        ERROR ("stenographer plugin: Instance `%s': "
+                "No URL has been configured.",
+                st->name);
+        status = -1;
+    }
+
+    if (status == 0)
+    {
+        user_data_t ud;
+        char callback_name[3*DATA_MAX_NAME_LEN];
+
+        memset (&ud, 0, sizeof (ud));
+        ud.data = st;
+        ud.free_func = (void *) stenographer_free;
+
+        memset (callback_name, 0, sizeof (callback_name));
```

Comment:
You don't need to do this: `ssnprintf()` will make sure that the buffer is always null-terminated.